src/openrct2/ride/Vehicle.cpp

```diff
@@ -7259,48 +7255,57 @@ void Vehicle::UpdateAdditionalAnimation()
                             }();
                             int32_t directionIndex = sprite_direction >> 1;
                             auto offset = SteamParticleOffsets[typeIndex][directionIndex];
-                            steam_particle_create(x + offset.x, y + offset.y, z + offset.z);
+                            steam_particle_create(
+                                x + offset.x * vehicleEntry->steam_effect_translation[0] / STEAM_EFFECT_TRANSLATION_COEFFICIENT,
+                                y + offset.y * vehicleEntry->steam_effect_translation[0] / STEAM_EFFECT_TRANSLATION_COEFFICIENT,
+                                z
+                                    + offset.z * vehicleEntry->steam_effect_translation[1]
+                                        / STEAM_EFFECT_TRANSLATION_COEFFICIENT);
                         }
                     }
                 }
                 Invalidate();
             }
             break;
-        case VEHICLE_ENTRY_ANIMATION_SWAN: // loc_6D6424
-            *curVar_C8 += _vehicleVelocityF64E08;
-            al = (*curVar_C8 >> 18) & 2;
-            if (animation_frame != al)
+        case VEHICLE_ENTRY_ANIMATION_SWAN:        // loc_6D6424
+            *curVar_C8 += _vehicleVelocityF64E08; // 2 frame animation, with frames at 0 and 2, otherwise identical to water
+                                                  // tricycles
+            // the frame at 2 means the exponent is -1 from water tricycles for the same playback speed
+            currentFrame = ((*curVar_C8 * vehicleEntry->animation_speed_multiplier) >> vehicleEntry->animation_speed_exponent)
+                & 2;
+            if (animation_frame != currentFrame)
             {
-                animation_frame = al;
+                animation_frame = currentFrame;
                 Invalidate();
             }
             break;
         case VEHICLE_ENTRY_ANIMATION_CANOES: // loc_6D6482
             *curVar_C8 += _vehicleVelocityF64E08;
-            eax = ((*curVar_C8 >> 13) & 0xFF) * 6;
-            ah = (eax >> 8) & 0xFF;
-            if (animation_frame != ah)
+            currentFrame = ((*curVar_C8 * vehicleEntry->animation_speed_multiplier) >> vehicleEntry->animation_speed_exponent)
+                % 6;
```

Comment:
This looks like a deviation from the old code - it used to do modulo 256, now it does modulo 6.

Reply:
It is actually the same but the original did it via some bit shifting